includes/enrolment/class-sensei-enrolment-provider-state-snapshot.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * File containing the class Sensei_Enrolment_Provider_State_Snapshot.
+ *
+ * @package sensei
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * A snapshot of the state of all enrolment providers.
+ */
+class Sensei_Enrolment_Provider_State_Snapshot implements JsonSerializable {
+
+	/**
+	 * Timestamp of the snapshot.
+	 *
+	 * @var int
+	 */
+	private $timestamp;
+
+	/**
+	 * An array of the status of all enrolment providers.
+	 *
+	 * [
+	 *      'provider_id' => [ 'enrolment_status' => true|false ]
+	 * ]
+	 *
+	 * @var array
+	 */
+	private $providers_status;
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param array   $providers_status The providers' status.
+	 * @param integer $timestamp        The timestamp of the snapshot.
+	 */
+	private function __construct( $providers_status, $timestamp = null ) {
+		$this->timestamp        = null === $timestamp ? time() : $timestamp;
+		$this->providers_status = $providers_status;
+	}
+
+	/**
+	 * Restore a snapshot from a JSON string.
+	 *
+	 * @param array $data Serialized state of object.
+	 *
+	 * @return self|false
+	 */
+	public static function from_serialized_array( $data ) {
+
+		if ( empty( $data ) ) {
+			return false;
+		}
+
+		return new self( $data['p'], $data['t'] );
+	}
+
+	/**
+	 * Create an empty snapshot.
+	 *
+	 * @param array $providers_status The providers' status.
+	 *
+	 * @return self
+	 */
+	public static function create( $providers_status = [] ) {
+		return new self( $providers_status );
+	}
+
+	/**
+	 * Create a duplicate of a snapshot.
+	 *
+	 * @param Sensei_Enrolment_Provider_State_Snapshot $other The snapshot to copy.
+	 *
+	 * @return self
+	 */
+	public static function duplicate( Sensei_Enrolment_Provider_State_Snapshot $other ) {
+		return new self( $other->providers_status );
```

Comment:
TIL [objects of same type](https://www.php.net/manual/en/language.oop5.visibility.php#language.oop5.visibility-other-objects) could access private properties.